Item controller 2

cdpiazza opened the pull request:
Agregada al capa DAO

Commit:
Cambios a ser confirmados:
	copiado:        src/main/java/com/app/rest/service/ItemService.java -> src/main/java/com/app/rest/dao/ItemDAO.java
	modificado:     src/main/java/com/app/rest/format/DateFormat.java
	modificado:     src/main/java/com/app/rest/service/ItemService.java

Commit:
Cambios a ser confirmados:
	modificado:     src/main/java/com/app/rest/service/ItemService.java

Commit:
Cambio en los ItemTypes a FOOD,SERVICE,OTHER:
	modificado:     src/main/java/com/app/rest/model/dto/ItemType.java
	modificado:     src/main/resources/import.sql
	modificado:     src/test/java/com/app/rest/test/service/ItemTest.java

Commit:
Cambios a ser confirmados:
	renombrado:     src/main/java/com/app/rest/dao/ItemDAO.java -> src/main/java/com/app/rest/dao/impl/ItemDAOHibernateImpl.java
	nuevo archivo:  src/main/java/com/app/rest/dao/interfaces/ItemDAO.java
	modificado:     src/main/java/com/app/rest/format/DateFormat.java
	modificado:     src/main/java/com/app/rest/model/dto/ItemDTO.java
	nuevo archivo:  src/main/java/com/app/rest/model/dto/ItemValidator.java
	modificado:     src/main/java/com/app/rest/service/ItemService.java
	modificado:     src/test/java/com/app/rest/test/service/ItemTest.java

Commit:
Cambios a ser confirmados:
	modificado:     src/main/java/com/app/rest/controller/ItemController.java
	modificado:     src/main/java/com/app/rest/model/dto/ItemDTO.java
	modificado:     src/main/java/com/app/rest/model/dto/ItemValidator.java
	modificado:     src/test/java/com/app/rest/test/service/ItemTest.java

alangarciac commented on src/main/java/com/app/rest/dao/interfaces/ItemDAO.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.Optional;
+
+@Service
+public interface ItemDAO {
```
Esta interface podría llamarse `AbstractDAO` y que la implementen todos los DAOs de la app

cdpiazza replied:
oki

cdpiazza replied:
En realidad, ahi tenemos un tema porque la interfaz tiene que tener las excepciones que se van a levantar, y ahi dejaria de ser abstracto, porque item tiene las de item, order las de order, etc.

alangarciac replied:
En ese caso las exceptions deberían pasar a heredar de `RuntimeException`

cdpiazza replied:
Por claridad prefiero q sean interfaces separadas y no olvidarme de catchear excepciones q hacerlas runtime y comermela en algun lado, es para peor no te parece?

alangarciac commented on src/main/java/com/app/rest/dao/impl/ItemDAOHibernateImpl.java:
```diff
@@ -0,0 +1,151 @@
+        return (ItemDetail, cq, cb) -> cb.like(ItemDetail.get("description"), "%" + description + "%");
+    }
+
+    public ItemDTO findById(Long id) throws ItemNotFoundException, ItemException {
```
Faltarían los `@Override` cuando se implementa una interfaz, se coloca el `@Override` para denotar que se se está implementando el método de la interfaz

cdpiazza replied:
oki

alangarciac commented on src/main/java/com/app/rest/model/dto/ItemValidator.java:
```diff
@@ -0,0 +1,8 @@
+import com.app.rest.exception.ItemValidateException;
+
+public interface ItemValidator {
+    void validate() throws ItemValidateException, ItemTypeException;
```
podríamos utlizar la interfaz ya definida `Checkeable`, para que quede consistente todos los DTOs con la misma interfaz

cdpiazza replied:
igual que con la interfaz del dao, tenemos un tema con las posibles excepciones, quedan pegadas al tipo de DTO, no se puede generalizar... o al menos no se como facilmente.

alangarciac replied:
utilziando `Checkeable` la validéz de un item queda sujeta a el booleano de retorno, desde mi punto de vista queda mas claro y no tenes problemas de exceptios en method signature. El método isSupported() retorna true o false, no dispara exceptions. Si existen se manejan internamente en el isSupported y retornaría false.

cdpiazza replied:
Bueno eso me cambia todo el manejo :(

cdpiazza replied:
El tema es que devolver simplemente un true,false, no me sirve para el validate porque hay distintos tipos de errores.

Ahi veo q haciendo q tire Exception generico, puedo hacer la interfaz, fijate a ver q te parece hacerlo asi.

alangarciac commented on src/main/java/com/app/rest/service/ItemService.java:
```diff
@@ -1,132 +1,43 @@
-            LOGGER.error(message);
-            throw new ItemException(message);
-        }
+        return itemDAO.findById(id);
```
acá podrías hacer el try/catch de ItemException y loguear el error. De esta forma el logging para ItemException y manejo de la misma queda en el service.

cdpiazza replied:
oki!

cdpiazza replied:
pero no se deberian loguear los posibles problemas de DDBB en el dao q es el que se conecta con la BBDD??

cdpiazza replied:
o vos decis de loguear en cada lugar una cosa, en el dao los posibles persistance error y en el service otras cosa?

cdpiazza replied:
bueno lo resuelvo loggueando en el service, creo un tipo de excepcion nuevo para encapsular los persistance

alangarciac replied:
_el dao los posibles persistance error y en el service otras cosa?_ esto mismo, el DAO que loguee problemas de la base o de su manejador, en este caso Hibernate o la interfaz JPA

cdpiazza replied:
bueno entonces el service igual no loguea casi nada, porque que haya un error de tipo, o un error de notfound, no son errores logueables, quedan para debbug nada mas. los logueo con ese clasificador

Commit:
Added Functionality: search by type

 Cambios a ser confirmados:
	modificado:     src/main/java/com/app/rest/controller/ItemController.java
	modificado:     src/main/java/com/app/rest/dao/impl/ItemDAOHibernateImpl.java
	modificado:     src/main/java/com/app/rest/service/ItemService.java
	modificado:     src/test/java/com/app/rest/test/service/ItemTest.java

alangarciac reviewed (changes requested):
Quizás el DAO está teniendo toda la responsabilidad. Solo enmascaría las exceptions en el DAO y dejaría al service el logging catcheando las exceptions del tipo `ItemException*`

Re-utilizaría la interfaz Checkeable para que todos los DTOs queden consistentes con la misma interfaz.

Haría el AbstracDAO, ya que vamos a interfacear, para que quede genérico y lo utilicen todos los DAOs

Bien por la interfaz del DAO, buena iniciativa.

Avisame cualquier cosa, lo vemos!

Commit:
Cambios a ser confirmados:
	modificado:     src/main/java/com/app/rest/dao/impl/ItemDAOHibernateImpl.java
	modificado:     src/main/java/com/app/rest/service/ItemService.java
	modificado:     src/main/resources/import.sql
	modificado:     src/test/java/com/app/rest/test/service/ItemTest.java

Commit:
modificado:     src/main/java/com/app/rest/controller/ItemController.java
	modificado:     src/main/java/com/app/rest/dao/impl/ItemDAOHibernateImpl.java
	modificado:     src/main/java/com/app/rest/dao/interfaces/ItemDAO.java
	renombrado:     src/main/java/com/app/rest/exception/ItemException.java -> src/main/java/com/app/rest/exception/itemExceptions/ItemException.java
	renombrado:     src/main/java/com/app/rest/exception/ItemNotFoundException.java -> src/main/java/com/app/rest/exception/itemExceptions/ItemNotFoundException.java
	nuevo archivo:  src/main/java/com/app/rest/exception/itemExceptions/ItemPersistanceException.java
	renombrado:     src/main/java/com/app/rest/exception/ItemTypeException.java -> src/main/java/com/app/rest/exception/itemExceptions/ItemTypeException.java
	renombrado:     src/main/java/com/app/rest/exception/ItemValidateException.java -> src/main/java/com/app/rest/exception/itemExceptions/ItemValidateException.java
	modificado:     src/main/java/com/app/rest/model/dto/ItemDTO.java
	modificado:     src/main/java/com/app/rest/model/dto/ItemType.java
	modificado:     src/main/java/com/app/rest/model/dto/ItemValidator.java
	modificado:     src/main/java/com/app/rest/model/dto/OrderDTO.java
	modificado:     src/main/java/com/app/rest/service/ItemService.java
	modificado:     src/test/java/com/app/rest/test/service/ItemTest.java

Commit:
modificado:     src/main/java/com/app/rest/dao/impl/ItemDAOHibernateImpl.java
	modificado:     src/main/java/com/app/rest/exception/itemExceptions/ItemNotFoundException.java
	modificado:     src/main/java/com/app/rest/model/dto/Checkable.java
	modificado:     src/main/java/com/app/rest/model/dto/ItemDTO.java
	borrado:        src/main/java/com/app/rest/model/dto/ItemValidator.java
	modificado:     src/main/java/com/app/rest/model/dto/OrderDTO.java
	modificado:     src/main/java/com/app/rest/service/ItemService.java

Commit:
remove unused imports

alangarciac commented on src/main/java/com/app/rest/controller/ItemController.java:
```diff
@@ -79,5 +80,13 @@
         }
     }
-
+    @GetMapping("/items/{X_type}")
```
{type} quedaria mejor, era un ejemplo lo del trello jajaja

cdpiazza replied:
bueno, quiza hacia algo q se yo

alangarciac commented on src/main/java/com/app/rest/dao/interfaces/ItemDAO.java:
```diff
@@ -0,0 +1,24 @@
+
+import java.util.List;
+
+@Service
```
no hace falta esta anotación para la interface

cdpiazza replied:
qdo del copypaste

alangarciac commented on src/main/java/com/app/rest/exception/itemExceptions/ItemNotFoundException.java:
```diff
@@ -1,6 +1,8 @@
-public class ItemNotFoundException extends OrderException{
+import com.app.rest.exception.OrderException;
+
+public class ItemNotFoundException extends Exception {
```
podría extender de ItemException en lugar de extender de Exception

cdpiazza replied:
Se rompen los handlers cuando lo haces heredar de itemException porq lo toma como la misma excepcion en el catch, para esto hay q hacer una excepcion general y una excepcion padre, lo hago y de ultima despues lo borramos.

alangarciac commented on src/main/java/com/app/rest/exception/itemExceptions/ItemPersistanceException.java:
```diff
@@ -0,0 +1,19 @@
+package com.app.rest.exception.itemExceptions;
+
+public class ItemPersistanceException extends Exception {
```
lo mismo aca, podria extender de ItemException

alangarciac commented on src/main/java/com/app/rest/exception/itemExceptions/ItemPersistanceException.java:
```diff
@@ -0,0 +1,19 @@
+package com.app.rest.exception.itemExceptions;
+
+public class ItemPersistanceException extends Exception {
```
TYPO en Persist**a**nce > Persistence

cdpiazza replied:
Ups

alangarciac commented on src/main/java/com/app/rest/exception/itemExceptions/ItemTypeException.java:
```diff
@@ -1,4 +1,4 @@
-package com.app.rest.exception;
+package com.app.rest.exception.itemExceptions;
 
 public class ItemTypeException extends Exception {
```
podria extender de ItemException

cdpiazza replied:
creo q no cambia mucho, pero bueno como quieras

alangarciac replied:
queda mas ordenado

alangarciac commented on src/main/java/com/app/rest/model/dto/Checkable.java:
```diff
@@ -3,5 +3,7 @@
 public interface Checkable {
 
     boolean isSupported();
+    void validate() throws Exception;
```
Si vamos a ir por validate con exception, podría throwear un `ValidationException`. Eliminaría el  boolean isSupported();

cdpiazza replied:
aa me gusta esa de validation exception y listo, qda generico

alangarciac reviewed (commented):
algunos cambios, minimos

Commit:
Changes requested in code-review done:
	modificado:     src/main/java/com/app/rest/controller/ItemController.java
	modificado:     src/main/java/com/app/rest/dao/impl/ItemDAOHibernateImpl.java
	modificado:     src/main/java/com/app/rest/dao/interfaces/ItemDAO.java
	nuevo archivo:  src/main/java/com/app/rest/exception/ValidationException.java
	nuevo archivo:  src/main/java/com/app/rest/exception/itemExceptions/ItemGeneralException.java
	modificado:     src/main/java/com/app/rest/exception/itemExceptions/ItemNotFoundException.java
	borrado:        src/main/java/com/app/rest/exception/itemExceptions/ItemPersistanceException.java
	nuevo archivo:  src/main/java/com/app/rest/exception/itemExceptions/ItemPersistenceException.java
	modificado:     src/main/java/com/app/rest/exception/itemExceptions/ItemTypeException.java
	borrado:        src/main/java/com/app/rest/exception/itemExceptions/ItemValidateException.java
	modificado:     src/main/java/com/app/rest/model/dto/Checkable.java
	modificado:     src/main/java/com/app/rest/model/dto/ItemDTO.java
	modificado:     src/main/java/com/app/rest/service/ItemService.java
	modificado:     src/test/java/com/app/rest/test/service/ItemTest.java

alangarciac reviewed (approved):
impeca!
